drivers/mcp2515/candev_mcp2515.c

```diff
@@ -0,0 +1,582 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_mcp2515
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the CAN controller driver
+ *
+ * @author      Toon Stegen <toon.stegen@altran.com>
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ * @}
+ */
+
+#include <string.h>
+#include <errno.h>
+#include <limits.h>
+#include <can/common.h>
+
+#include "candev_mcp2515.h"
+#include "can/common.h"
+#include "can/device.h"
+#include "mcp2515.h"
+#include "periph_conf.h"
+#include "thread.h"
+#include "sched.h"
+#include "mutex.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define MAILBOX_USED 1
+#define MAILBOX_EMPTY 0
+
+#define MAILBOX_TO_INTR(mailbox) (INT_RX0 + (mailbox))
+
+#ifndef CANDEV_MCP2515_DEFAULT_BITRATE
+#define CANDEV_MCP2515_DEFAULT_BITRATE 500000
+#endif
+
+#ifndef CANDEV_MCP2515_DEFAULT_SPT
+#define CANDEV_MCP2515_DEFAULT_SPT 875
+#endif
+
+static int _init(candev_t *candev);
+static int _send(candev_t *candev, const struct can_frame *frame);
+static void _isr(candev_t *candev);
+static int _set(candev_t *candev, canopt_t opt, void *value, size_t value_len);
+static int _get(candev_t *candev, canopt_t opt, void *value, size_t max_len);
+static int _abort(candev_t *candev, const struct can_frame *frame);
+static int _set_filter(candev_t *dev, const struct can_filter *filter);
+static int _remove_filter(candev_t *dev, const struct can_filter *filter);
+
+static void _irq_rx(candev_mcp2515_t *dev, int handle);
+static void _irq_tx(candev_mcp2515_t *dev, int handle);
+static void _irq_error(candev_mcp2515_t *dev);
+static void _irq_message_error(candev_mcp2515_t *dev);
+static void _irq_wakeup(const candev_mcp2515_t *dev);
+static void _send_event(const candev_mcp2515_t *dev, candev_event_t event,
+                        void *arg);
+
+static const candev_driver_t candev_mcp2515_driver = {
+    .send = _send,
+    .init = _init,
+    .isr = _isr,
+    .get = _get,
+    .set = _set,
+    .abort = _abort,
+    .set_filter = _set_filter,
+    .remove_filter = _remove_filter,
+};
+
+static const struct can_bittiming_const bittiming_const = {
+    .tseg1_min = 3,
+    .tseg1_max = 16,
+    .tseg2_min = 2,
+    .tseg2_max = 8,
+    .sjw_max = 4,
+    .brp_min = 1,
+    .brp_max = 64,
+    .brp_inc = 1,
+};
+
+static inline int _max_filters(int mailbox)
+{
+    return mailbox == 0 ? MCP2515_FILTERS_MB0 : MCP2515_FILTERS_MB1;
+}
+
+void candev_mcp2515_init(candev_mcp2515_t *dev,
+                         const candev_mcp2515_conf_t *conf)
+{
+    memset(dev, 0, sizeof(*dev));
+    dev->candev.driver = &candev_mcp2515_driver;
+
+    struct can_bittiming timing = { .bitrate = CANDEV_MCP2515_DEFAULT_BITRATE,
+                                    .sample_point =
+                                        CANDEV_MCP2515_DEFAULT_SPT };
+    can_device_calc_bittiming(conf->clk / 2, &bittiming_const, &timing);
```

Comment:
I guess we could, but I personally don't think that's a good idea. The clockrate is a hardware-dependent value. To me it feels that using only half of the hardware clockrate in code can only lead to confusion. 
The compiler will probably optimize this division anyways.

Reply:
Been a while since I worked on this. Took a while to investigate this, and I'm still not 100% sure. I think this is the reasoning behind it:

```
Tq = 2*BRP*Tosc
(with BRP = 1, because the BRP register is not set and BRP = BRP[5:0] + 1)
Tq = 2*Tosc
1/fq = 2/fosc
fq = fosc/2
```

Tq is the time Time Quantum that make up the bit time. There's a timing diagram on [p40 of the datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/MCP2515-Stand-Alone-CAN-Controller-with-SPI-20001801J.pdf) that shows how the frequency is halved. Above it is some explanation. 

But again, I'm not entirely sure if this is the correct explanation for the /2. @vincent-d can you maybe confirm this?

Reply:
Ah yes thanks for the pointer, the reference you point out makes me think the same, could you add a comment in code for this?